src/bitcoin/P2tr.mo

```diff
@@ -30,4 +43,77 @@ module {
             };
         };
     };
+
+    // Create script for the given P2TR key spend address.
+    public func leafScript(bip340_spender_public_key : [Nat8]) : Result.Result<Script, Text> {
+        if (bip340_spender_public_key.size() != 32) {
+            return #err("Invalid BIP-340 public key length: expected 32 but got " # Nat.toText(bip340_spender_public_key.size()));
+        };
+        #ok([
+            // #opcode(#OP_PUSHBYTES_32) is implicit and added by the
+            // #data below
+            #data(bip340_spender_public_key),
+            #opcode(#OP_CHECKSIG),
+        ]);
+    };
+
+    public func leafHash(leaf_script : Script.Script) : [Nat8] {
+        // BIP-342 tapscript
+        let TAPROOT_LEAF_TAPSCRIPT : [Nat8] = [0xc0];
+        let script_bytes = Script.toBytes(leaf_script);
+        Hash.taggedHash(Array.flatten([TAPROOT_LEAF_TAPSCRIPT, script_bytes]), "TapLeaf");
+    };
+
+    public func tweakFromKeyAndHash(internal_key : [Nat8], hash : [Nat8]) : Result.Result<Fp.Fp, Text> {
```

Comment:
Does this correspond to the subpart
```
    t = int_from_bytes(tagged_hash("TapTweak", pubkey + h))
    if t >= SECP256K1_ORDER:
        raise ValueError
```
of [taproot_tweak_pubkey](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs)?

Reply:
yes, added a comment